### PR TITLE
Optimize preconnect callbacks

### LINKED ISSUE #22975: Preconnect timer and BUILT event delays FCP
Noticed this while futzing around with IntersectionObserver instrumentation. On a test page with ~400 amp-img, the two combined take about 14% of pre-FCP script evaluation time. 

1. The `setTimeout` in preconnect results in O(N) timer installations in browser ([Install Timer](https://developers.google.com/web/tools/chrome-devtools/evaluate-performance/performance-reference)).
2. `AmpEvents.BUILT` is dispatched O(N) times but only has one listener in `amp-fx-flying-carpet.js`.

We should batch the preconnect timer and probably remove `AmpEvents.BUILT` outright.

[1] https://github.com/ampproject/amphtml/blob/cd8ce972f9be9d36ea1574cb8347be9645b327dd/src/custom-element.js#L574-L581

[2] https://github.com/ampproject/amphtml/blob/cd8ce972f9be9d36ea1574cb8347be9645b327dd/src/service/resource.js#L350

Ran on Chrome Mac Pro 2013 with 4x CPU throttling:

Screen Shot 2019-06-24 at 10 55 33 AM

/cc @ampproject/wg-performance @ampproject/wg-runtime

### PR DESCRIPTION
**Pull request by Enriqe**
Closes #22975

Before:
image

After (using `startupChunk`):
image

### CODE REVIEW

File: src/custom-element.js [hunk 1]
```diff
@@ -573,29 +573,32 @@ function createBaseCustomElementClass(win) {
...
     /**
      * Called to instruct the element to preconnect to hosts it uses during
-     * layout.
+     * layout. If attempting an early preconnect, it requests to preconnect
+     * to resources beforehand.
      * @param {boolean} onLayout Whether this was called after a layout.
      * @this {!Element}
      */
-    preconnect(onLayout) {
+    maybePreconnect(onLayout) {
```
Comment (jridgewell) [inline src/custom-element.js]: I don't understand why this was split into two methods?

File: src/custom-element.js [hunk 2]
```diff
@@ -573,29 +573,32 @@ function createBaseCustomElementClass(win) {
...
+      }
+    }
+
+    /**
+     * Preconnect to host.
+     */
+    preconnect() {
+      const TAG = this.tagName;
+      if (!this.ownerDocument) {
+        dev().error(TAG, 'preconnect without ownerDocument');
```
Comment (jridgewell) [inline src/custom-element.js]: This is a change in behavior. The element's ownerDocument could have been removed during the delay, not immediately.

File: src/custom-element.js [hunk 3]
```diff
@@ -573,29 +573,32 @@ function createBaseCustomElementClass(win) {
...
-          this.implementation_.preconnectCallback(onLayout);
-        }, 1);
+        this.getResources().requestPreconnect(this);
+      }
+    }
+
+    /**
+     * Preconnect to host.
+     */
+    preconnect() {
```
Comment (jridgewell) [inline src/custom-element.js]: This used to take the `onLayout` boolean.

File: src/custom-element.js [hunk 4]
```diff
@@ -535,7 +535,7 @@ function createBaseCustomElementClass(win) {
         }
       }).then(
         () => {
-          this.preconnect(/* onLayout */ false);
+          this.maybePreconnect(/* onLayout */ false);
```
Comment (jridgewell) [inline src/custom-element.js]: Should this be private?

File: src/service/resources-impl.js [hunk 5]
```diff
@@ -497,6 +498,18 @@ export class Resources {
     /** @private {?Array<!Resource>} */
     this.pendingBuildResources_ = [];
 
+    /** @private {?Array<!Resource>} */
+    this.pendingPreconnectResources_ = [];
+
+    // If we do early preconnects we delay them a bit. This is kind of
+    // an unfortunate trade off, but it seems faster, because the DOM
+    // operations themselves are not free and might delay
+    this.eventuallyCallPreconnects_ = debounce(
```
Comment (jridgewell) [inline src/service/resources-impl.js]: `debounce` waits X ms after the latest call before doing anything, which means they could be waiting longer than 5ms to actually be called.

File: src/service/resources-impl.js [hunk 6]
```diff
@@ -912,6 +925,29 @@ export class Resources {
...
+  /**
+   * Calls resource manager's preconnect and removes resource from the pending
+   * buffer.
+   * @private
+   */
+  callPreconnects_() {
+    for (let i = 0; i < this.pendingPreconnectResources_.length; i++) {
+      const resource = this.pendingPreconnectResources_[i];
+      resource.preconnect();
+      this.pendingPreconnectResources_.splice(i, 1);
```
Comment (jridgewell) [inline src/service/resources-impl.js]: This makes `callPreconnects` an n<sup>2</sup> algorithm

File: src/service/resources-impl.js [hunk 7]
```diff
@@ -912,6 +925,29 @@ export class Resources {
     dev().fine(TAG_, 'element upgraded:', resource.debugid);
   }
 
+  /**
+   * Schedule resources for preconnecting.
+   * @param {!Element} element
+   */
+  requestPreconnect(element) {
```
Comment (jridgewell) [inline src/service/resources-impl.js]: I feel like we can do a better optimization by applying this at the `Timer` level and not on the `preconnect` method specifically.

File: extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js [hunk 8]
```diff
@@ -102,7 +101,13 @@ export class AmpFlyingCarpet extends AMP.BaseElement {
     });
     if (this.firstLayoutCompleted_) {
       this.scheduleLayout(this.children_);
-      listen(this.element, AmpEvents.BUILT, this.layoutBuiltChild_.bind(this));
+      this.children_.forEach(child => {
```
Comment (jridgewell) [inline extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js]: This isn't equivalent. We're listening for new children to be added to the flying carpet, not for already know children to upgrade. "Necessary since not all children will be built by the time the flying-carpet has its #layoutCallback called."

Let's split this into a separate PR.

Comment (cramforce) [reply extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js]: Added possibly more complete impl. to #23767

File: src/service/timer-impl.js [hunk 9]
```diff
@@ -78,13 +81,44 @@ export class Timer {
...
+   * Calls previously batched callbacks for a given ms key.
+   * @param {number} delay
+   * @private
+   */
+  callCallbacksForDelay_(delay) {
+    const callbacks = this.delayQueue_[delay];
+    delete this.delayQueue_[delay];
+
+    for (let i = 0; i < callbacks.length; i++) {
+      const cb = callbacks[i];
```
Comment (jridgewell) [inline src/service/timer-impl.js]: This needs to be wrapped in a try-catch, since any `cb` could throw.

File: src/service/timer-impl.js [hunk 10]
```diff
@@ -78,13 +81,44 @@ export class Timer {
...
         reportError(e);
         throw e;
       }
     };
-    return this.win.setTimeout(wrapped, opt_delay);
+
+    const delayEnd = Date.now() + opt_delay;
+    if (!this.delayQueue_[delayEnd]) {
+      this.delayQueue_[delayEnd] = [];
+      id = this.win.setTimeout(
```
Comment (jridgewell) [inline src/service/timer-impl.js]: We can't return this id, since if the user calls `cancelTimeout` with it it would prevent all `cb`s registered for the same time from firing.

File: src/service/timer-impl.js (hunk 9 above)
Comment (Enriqe) [reply src/service/timer-impl.js]: This cb is already wrapped in a try-catch before being pushed to the queue, or am I missing something?

File: src/service/timer-impl.js (hunk 10 above)
Comment (Enriqe) [reply src/service/timer-impl.js]: Oh right

**Comment by jridgewell** [comment]
> because it relies on the calls coming in the same milli second

Which is the use case from https://github.com/ampproject/amphtml/issues/22975. But it's not just in the same ms, it's scheduled to run at the same ms. So `time = 100, delay = 10` and `time = 101, delay = 9` will share a timeout now.

> and has the additional downside that individual callbacks are no longer scheduled as tasks.

What's the downside here? It'll require a try-catch, but this should work fine.

**Comment by cramforce** [comment]
Downside is that e.g. user actions cannot interrupt execution.

**Comment by Enriqe** [comment]
Ready for another review 👍
